File: bench/prefix_scan.c
// benchmark-contract: prefix-scan;seed=123456789;batches=1000000;width=16;value-mask=65535
#include <stdint.h>
#include <stdio.h>

static inline void verify_checksum(uint64_t value) {
#ifdef BENCH_VERIFY
  fwrite(&value, sizeof(value), 1, stdout);
#else
  (void)value;
#endif
}

int main(void) {
  const uint64_t batches = 1000000ULL;
  uint64_t values[16] = {0};
  uint64_t state = 123456789ULL;
  uint64_t checksum = 0;
  uint64_t batch = 0;

  while (batch < batches) {
    uint64_t index = 0;
    while (index < 16ULL) {
      state = (state * 1664525ULL + 1013904223ULL) & 0xffffffffULL;
      values[index] = state & 0xffffULL;
      ++index;
    }

    index = 1;
    while (index < 16ULL) {
      values[index] += values[index - 1ULL];
      ++index;
    }
    checksum ^= values[15];
    ++batch;
  }

  uint64_t result = state ^ checksum;
  verify_checksum(result);
  return (int)(result & 0x7fULL);
}

File: bench/prefix_scan.rs
// benchmark-contract: prefix-scan;seed=123456789;batches=1000000;width=16;value-mask=65535
const BATCHES: u64 = 1_000_000;
const SEED: u64 = 123_456_789;

fn finish(value: u64) -> ! {
    // See the C peer: one decimal line, masked to 63 bits.
    println!("{}", value & 0x7fff_ffff_ffff_ffff);
    std::process::exit(0)
}

fn main() {
    let mut values = [0u64; 16];
    let mut state = SEED;
    let mut checksum = 0u64;
    let mut batch = 0u64;

    while batch < BATCHES {
        let mut index = 0usize;
        while index < 16 {
            state = state.wrapping_mul(1_664_525).wrapping_add(1_013_904_223) & 0xffff_ffff;
            values[index] = state & 0xffff;
            index += 1;
        }

        index = 1;
        while index < 16 {
            values[index] = values[index].wrapping_add(values[index - 1]);
            index += 1;
        }
        checksum ^= values[15];
        batch += 1;
    }

    finish(state ^ checksum);
}
